stdlib/public/runtime/MetadataLookup.cpp

```diff
@@ -54,17 +54,106 @@ static Demangler getDemanglerForRuntimeTypeResolution() {
   // the type context descriptor, so that if we see a symbolic reference in the
   // mangled name we can immediately find the associated metadata.
-  dem.setSymbolicReferenceResolver([&](int32_t offset,
-                                       const void *base) -> NodePointer {
-    auto absolute_addr = (uintptr_t)detail::applyRelativeOffset(base, offset);
-    auto reference = dem.createNode(Node::Kind::SymbolicReference, absolute_addr);
-    auto type = dem.createNode(Node::Kind::Type);
-    type->addChild(reference, dem);
-    return type;
-  });
+  dem.setSymbolicReferenceResolver(ResolveAsSymbolicReference(dem));
   return dem;
 }
 
+NodePointer
+ResolveAsSymbolicReference::operator()(SymbolicReferenceKind kind,
+                                       Directness isIndirect,
+                                       int32_t offset,
+                                       const void *base) {
+  // Resolve the absolute pointer to the entity being referenced.
+  auto ptr = detail::applyRelativeOffset(base, offset);
+  if (isIndirect == Directness::Indirect) {
+    ptr = *(const uintptr_t *)ptr;
+  }
+
+  // Figure out this symbolic reference's grammatical role.
+  Node::Kind nodeKind;
+  switch (kind) {
+  case Demangle::SymbolicReferenceKind::Context: {
+    auto descriptor = (const ContextDescriptor *)ptr;
+    switch (descriptor->getKind()) {
+    case ContextDescriptorKind::Protocol:
+      nodeKind = Node::Kind::ProtocolSymbolicReference;
+      break;
+      
+    default:
+      if (auto typeContext = dyn_cast<TypeContextDescriptor>(descriptor)) {
+        nodeKind = Node::Kind::TypeSymbolicReference;
+        break;
+      }
+      
+      // References to other kinds of context aren't yet implemented.
+      return nullptr;
+    }
+    break;
+  }
+    
+  case Demangle::SymbolicReferenceKind::ProtocolConformance: {
+    nodeKind = Node::Kind::ProtocolConformanceSymbolicReference;
+    break;
+  }
+  }
+  
+  return Dem.createNode(nodeKind, ptr);
+}
+
+static NodePointer
+_buildDemanglingForSymbolicReference(SymbolicReferenceKind kind,
+                                     const void *resolvedReference,
+                                     Demangler &Dem) {
+  switch (kind) {
+  case SymbolicReferenceKind::Context:
+    return _buildDemanglingForContext(
+      (const ContextDescriptor *)resolvedReference, {}, Dem);
+    
+  case SymbolicReferenceKind::ProtocolConformance: {
+    swift_runtime_unreachable("not implemented");
+    /* TODO
+    auto conformance = (const ProtocolConformanceDescriptor *)resolvedReference;
+    
+    auto protocolMangling =
+      _buildDemanglingForContext(conformance->getProtocol(), {}, Dem);
+    
+    const ModuleContextDescriptor *module;
+    if (auto retroContext = conformance->getRetroactiveContext()) {
+      module = retroContext->getModuleContext();
+    } else {
+      // FIXME: Assume the protocol's module for a non-retroactive conformance
```

Comment:
Non-retroactive conformances could be either in the protocol's module or the conforming type's module. If we need it, we could add a bit to the conformance descriptor to say which one it is...